drgn/helpers/linux/slab.py

```diff
@@ -53,3 +55,163 @@ def print_slab_caches(prog: Program) -> None:
+) -> List[Object]:
+    """
+    Get the list of all allocated objects of a given slab and type
+
+    >>> dentry_slab = find_slab_cache(prog, "dentry")
+    >>> dentry_type = prog.type("struct dentry")
+    >>> get_allocated_slab_objects(prog, dentry_slab, dentry_type)
+    [Object(prog, 'struct dentry', address=0xffff888c8417de00), Object(prog, 'struct dentry', address=0xffff888c8417dec0)]
+
+    :param slab_cache: ``struct kmem_cache``
+    :param obj_type: type of object contained in the slab
+    """
+    # Linux kernel commit d122019bf061cccc4583eb9ad40bf58c2fe517be ("mm: Split slab into its own type")
+    # (in v5.17) moved slab fields out of struct page into struct slab and allows page to be cast to slab
+    slub = False
+    try:
+        freelist_index_size = sizeof(prog.type("freelist_idx_t"))
+    except LookupError:
+        slub = True
+
+    red_left_pad = 0
+    if prog.type("struct kmem_cache").has_member("red_left_pad"):
+        red_left_pad += slab_cache.red_left_pad.value_()
+
+    def _slub_page_objects(
+        prog: Program, slab_cache: Object, page_or_slab: Object, obj_type: Type
+    ) -> List[Object]:
+        """
+        Get the list of all objects from a slub page
+        :param slab_cache: ``struct kmem_cache``
+        :param page_or_slab: ``struct page`` or ``struct slab`` depending on the kernel version
+        :param obj_type: type of object contained in the slab_cache
+        """
+        addr = page_to_virt(cast(prog.type("struct page *"), page_or_slab)).value_()
+        if not isinstance(addr, int):
+            addr = addr.value_()
+        addr += red_left_pad
+
+        end = addr + slab_cache.size * page_or_slab.objects
+        ret = []
+
+        slab_cache_size = slab_cache.size.value_()
+        freelist_set = obtain_freelist_slub(page_or_slab, slab_cache.offset.value_())
+        while addr < end:
+            if addr not in freelist_set:
+                ret.append(Object(prog, obj_type, address=addr))
+            addr += slab_cache_size
+        return ret
+
+    def _slab_page_objects(
+        prog: Program, slab_cache: Object, page_or_slab: Object, obj_type: Type
+    ) -> List[Object]:
+        """
+        Get the list of all objects from a slab_cache page.
+        Page may be cast to slab for v5.17 and on
+        :param slab_cache: ``struct kmem_cache``
+        :param page_or_slab: ``struct page`` or ``struct slab`` depending on the kernel version
+        :param obj_type: type of object contained in the slab_cache
+        """
+        freelist_set = obtain_freelist_slab(prog, page_or_slab, freelist_index_size)
+        ret = []
+        offset = 0
+        if prog.type("struct kmem_cache").has_member("obj_offset"):
+            offset = slab_cache.obj_offset.value_()
+
+        for i in range(0, slab_cache.num):
+            if i in freelist_set:
+                continue
+            addr = page_or_slab.s_mem.value_() + i * slab_cache.size + offset
```

Comment:
Get `page_or_slab.s_mem.value_()` once outside of the loop, and get `slab_cache.size` once outside of the function (which can be shared with the SLUB code path).